### PR TITLE
Enable fluid ads

### PR DESCRIPTION
**Pull request by regiskuckaertz**
This change will enable fluid ads for the top banner only. It is hidden behind the `FluidAdverts` switch.
### SASS

I created a `.ad-slot--fluid` modifier but because currently top banner ads have a fixed width and left padding on desktop, I had to compound classes to cancel it. There may be a more idiomatic way of doing this at the Guardian though.

### CODE REVIEW

**Comment by johnduffell** [comment]
not sure what fluid ads are, but is this likely to cause movement of the page after it's loaded?

**Comment by regiskuckaertz** [comment]
@johnduffell it is meant to be fluid in the sense that it occupies 100% of the viewport width—the same height is kept throughout to prevent layout thrashing. At the moment though, there is a nasty 'appear' effect that animates on the height and I'm still trying to figure out where it comes from.

File: common/app/conf/switches/CommercialSwitches.scala [hunk 1]
```diff
@@ -32,6 +32,15 @@ trait CommercialSwitches {
     exposeClientSide = true
   )
 
+  val FluidAdvertsSwitch = Switch(
+    "Commercial",
+    "fluid-adverts",
+    "Enable fluid ads",
```
Comment (johnduffell) [inline common/app/conf/switches/CommercialSwitches.scala]: might be worth putting a bt more explanation here

**Comment by johnduffell** [comment]
sounds good, jump (and jank) are always painful for users! :+1: although I don't know much about the dfp-api.js code

File: common/app/conf/switches/CommercialSwitches.scala [hunk 2]
```diff
@@ -32,6 +32,15 @@ trait CommercialSwitches {
...
+    safeState = Off,
+    sellByDate = never,
```
Comment (jbreckmckye) [inline common/app/conf/switches/CommercialSwitches.scala]: Never? Never ever... ever?

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js [hunk 3]
```diff
@@ -404,24 +406,28 @@ define([
...
+                slot,
+                size,
+                sizeMapping;
+
+            if ($adSlot.data('out-of-page')) {
+                slot = googletag.defineOutOfPageSlot(adUnit, id);
+            } else if ($adSlot.data('fluid') && cookies.get('adtest') === 'tm2') {
+                $adSlot.addClass('ad-slot--fluid');
+                sizeMapping = defineSlotSizes($adSlot);
+                // regis kuckaertz (Nov 30) – SizeMappingBuilder does not handle 'fluid' very well,
```
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: You don't need to annotate your name and date of authorship - we already have that information in `git blame`.

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js [hunk 4]
```diff
@@ -404,24 +406,28 @@ define([
                 adUnit         = adUnitOverride ?
                     ['/', config.page.dfpAccountId, '/', adUnitOverride].join('') : config.page.adUnit,
                 id             = $adSlot.attr('id'),
-                sizeMapping    = defineSlotSizes($adSlot),
+                slot,
+                size,
+                sizeMapping;
+
+            if ($adSlot.data('out-of-page')) {
+                slot = googletag.defineOutOfPageSlot(adUnit, id);
```
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: Does this branch miss out on the defineSizeMapping step?

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js [hunk 5]
```diff
@@ -404,24 +406,28 @@ define([
...
+            } else if ($adSlot.data('fluid') && cookies.get('adtest') === 'tm2') {
```
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: We're not going to have adtest cookies on for the most part in production - is this just for internal demonstrations? Is this actually going to be released to production?

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js [hunk 6]
```diff
@@ -404,24 +406,28 @@ define([
...
+                $adSlot.addClass('ad-slot--fluid');
+                sizeMapping = defineSlotSizes($adSlot);
+                // regis kuckaertz (Nov 30) – SizeMappingBuilder does not handle 'fluid' very well,
+                // so instead we add it manually ourselves to the end of each array of sizes
```
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: So are we circumventing the normal API here? Are we actually supposed to manipulate these sizeMap objects?

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js [hunk 7]
```diff
@@ -404,24 +406,28 @@ define([
...
-                slot = (
-                    $adSlot.data('out-of-page') ?
-                        googletag.defineOutOfPageSlot(adUnit, id) :
-                        googletag.defineSlot(adUnit, size, id)
-                    )
-                    .addService(googletag.pubads())
-                    .defineSizeMapping(sizeMapping)
-                    .setTargeting('slot', slotTarget);
+                size = uniq(
+                    flatten(sizeMapping, true, function (map) { return map[1]; }),
```
Comment (jbreckmckye) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: We don't normally collapse lambdas.

File: common/app/conf/switches/CommercialSwitches.scala (hunk 1 above)
Comment (regiskuckaertz) [inline common/app/conf/switches/CommercialSwitches.scala]: good call, I'm on it.

**Comment by regiskuckaertz** [comment]
Thanks!

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js (hunk 5 above)
Comment (regiskuckaertz) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: this is currently a beta feature of DFP and we are testing it.

File: static/src/javascripts/projects/common/modules/commercial/dfp-api.js (hunk 6 above)
Comment (regiskuckaertz) [inline static/src/javascripts/projects/common/modules/commercial/dfp-api.js]: SizeMappingBuilder is just a helper, but I'm not circumventing it. Instead, I'm building on it by injecting the 'fluid' size at the end.

File: common/app/conf/switches/CommercialSwitches.scala (hunk 2 above)
Comment (regiskuckaertz) [inline common/app/conf/switches/CommercialSwitches.scala]: good call, I'll set a date in the near future.

**Pull request merged**